### ISSUE TITLE
Nuget add hangs on CentOS

### ISSUE THREAD

**Issue by balneaves** [NuGet/Home#2708]
I've been running the latest dotnet core builds via Jenkins on our Centos build server, however the last step of KoreBuild always hangs, which is when it tries to run 'nuget add'.  All other commands (restore, etc) seem to work just fine.

I can reproduce this manually with any .nupkg, below is with one of the standard .net libs:

```
mono ~/NuGet.exe add System.IO.4.1.0-rc2-23826.nupkg -source ~/test -verbosity detailed
Acquiring lock for the installation of System.IO 4.1.0-rc2-23826
Acquired lock for the installation of System.IO 4.1.0-rc2-23826
Installing System.IO 4.1.0-rc2-23826.
```

It hangs at that point, and unfortunately can't seem to get any other information to diagnose the issue.  It doesnt throw an error, and I can't Ctrl+C.  Have to kill -9 the process.

It seems to create the temp file for the copy, which is the right size and am able to unzip the temp file so the copy is complete:
`-rw-rw-r-- 1 centos centos 519K May  3 18:56 /home/centos/test/system.io/4.1.0-rc2-23826/95gik1i7.62j`

Versions:

```
NuGet Version: 3.5.0.1240
Mono JIT compiler version 4.2.3 (Stable 4.2.3.4/832de4b Tue Mar 15 11:39:54 EDT 2016)
CentOS Linux release 7.2.1511 (Core)
```

**Comment by rrelyea** [comment]
plan is to enable "dotnet nuget add ..." in future and run directly on core clr.
Will get back to the mono issue at a later time.

**Comment by migueldeicaza** [comment]
From a different terminal, find out the PID for the process running NuGet, and then type:

kill -QUIT PID

This will produce a stack trace that shows where the Mono process is hanging.

**Comment by balneaves** [comment]
```
Full thread dump:

"Threadpool worker"  at <unknown> <0xffffffff>
  at (wrapper managed-to-native) object.__icall_wrapper_mono_delegate_end_invoke (object,intptr) <0xffffffff>
  at (wrapper delegate-end-invoke) <Module>.end_invoke_int__this___IAsyncResult (System.IAsyncResult) <0xffffffff>
  at System.IO.FileStream.EndRead (System.IAsyncResult) <0x000bc>
  at System.IO.FileStream.Read (byte[],int,int) <0x000d5>
  at SharpCompress.IO.NonDisposingStream.Read (byte[],int,int) <0x0003e>
  at System.IO.BinaryReader.FillBuffer (int) <0x000ec>
  at System.IO.BinaryReader.ReadUInt32 () <0x00015>
  at SharpCompress.Common.Zip.SeekableZipHeaderFactory/<ReadSeekableHeader>c__Iterator0.MoveNext () <0x00128>
  at SharpCompress.Archive.Zip.ZipArchive/<LoadEntries>c__Iterator0.MoveNext () <0x0048c>
  at SharpCompress.LazyReadOnlyCollection`1/LazyLoader<T_REF>.MoveNext () <0x000c2>
  at System.IO.Compression.ZipArchive.CreateZip (System.IO.Stream,System.IO.Compression.ZipArchiveMode) <0x00308>
  at System.IO.Compression.ZipArchive..ctor (System.IO.Stream,System.IO.Compression.ZipArchiveMode,bool) <0x0007f>
  at NuGet.Packaging.PackageArchiveReader..ctor (System.IO.Stream,bool,NuGet.Frameworks.IFrameworkNameProvider,NuGet.Frameworks.IFrameworkCompatibilityProvider) <0x00053>
  at NuGet.Packaging.PackageArchiveReader..ctor (System.IO.Stream) <0x0004b>
  at NuGet.Packaging.PackageExtractor/<>c__DisplayClass2_0/<<InstallFromSourceAsync>b__0>d.MoveNext () <0x007a3>
  at System.Runtime.CompilerServices.AsyncMethodBuilderCore/MoveNextRunner.InvokeMoveNext (object) <0x000bf>
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext,System.Threading.ContextCallback,object,bool) <0x001c6>
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext,System.Threading.ContextCallback,object,bool) <0x00020>
  at System.Runtime.CompilerServices.AsyncMethodBuilderCore/MoveNextRunner.Run () <0x000ce>
  at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction (System.Action,bool,System.Threading.Tasks.Task&) <0x00071>
  at System.Threading.Tasks.Task.FinishContinuations () <0x00150>
  at System.Threading.Tasks.Task.FinishStageThree () <0x00068>
  at System.Threading.Tasks.Task`1<System.Threading.Tasks.VoidTaskResult>.TrySetResult (System.Threading.Tasks.VoidTaskResult) <0x000a3>
  at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<System.Threading.Tasks.VoidTaskResult>.SetResult (System.Threading.Tasks.VoidTaskResult) <0x0005e>
  at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<System.Threading.Tasks.VoidTaskResult>.SetResult (System.Threading.Tasks.Task`1<System.Threading.Tasks.VoidTaskResult>) <0x00053>
  at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetResult () <0x0001a>
  at System.IO.Stream/<CopyToAsyncInternal>c__async0.MoveNext () <0x00528>
  at System.Runtime.CompilerServices.AsyncMethodBuilderCore/MoveNextRunner.InvokeMoveNext (object) <0x000bf>
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext,System.Threading.ContextCallback,object,bool) <0x001c6>
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext,System.Threading.ContextCallback,object,bool) <0x00020>
  at System.Runtime.CompilerServices.AsyncMethodBuilderCore/MoveNextRunner.Run () <0x000ce>
  at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction (System.Action,bool,System.Threading.Tasks.Task&) <0x00071>
  at System.Threading.Tasks.Task.FinishContinuations () <0x00150>
  at System.Threading.Tasks.Task.FinishStageThree () <0x00068>
  at System.Threading.Tasks.Task`1<int>.TrySetResult (int) <0x00098>
  at System.Threading.Tasks.TaskFactory`1/FromAsyncTrimPromise`1<int, System.IO.Stream>.Complete (System.IO.Stream,System.Func`3<System.IO.Stream, System.IAsyncResult, int>,System.IAsyncResult,bool) <0x0007e>
  at System.Threading.Tasks.TaskFactory`1/FromAsyncTrimPromise`1<int, System.IO.Stream>.CompleteFromAsyncResult (System.IAsyncResult) <0x00196>
  at System.IO.Stream/ReadWriteTask.InvokeAsyncCallback (object) <0x00058>
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext,System.Threading.ContextCallback,object,bool) <0x001c6>
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext,System.Threading.ContextCallback,object,bool) <0x00020>
  at System.IO.Stream/ReadWriteTask.System.Threading.Tasks.ITaskCompletionAction.Invoke (System.Threading.Tasks.Task) <0x000f9>
  at System.Threading.Tasks.Task.FinishContinuations () <0x0020d>
  at System.Threading.Tasks.Task.FinishStageThree () <0x00068>
  at System.Threading.Tasks.Task.FinishStageTwo () <0x0009a>
  at System.Threading.Tasks.Task.Finish (bool) <0x00089>
  at System.Threading.Tasks.Task.ExecuteWithThreadLocal (System.Threading.Tasks.Task&) <0x00123>
  at System.Threading.Tasks.Task.ExecuteEntry (bool) <0x000f3>
  at System.Threading.Tasks.Task.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem () <0x0000e>
  at System.Threading.ThreadPoolWorkQueue.Dispatch () <0x001d6>
  at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback () <0x00008>
  at (wrapper runtime-invoke) <Module>.runtime_invoke_bool (object,intptr,intptr,intptr) <0xffffffff>

"<unnamed thread>"  at <unknown> <0xffffffff>
  at (wrapper managed-to-native) System.Threading.Monitor.Monitor_wait (object,int) <0xffffffff>
  at System.Threading.Monitor.Wait (object,int) <0x00042>
  at System.Threading.ManualResetEventSlim.Wait (int,System.Threading.CancellationToken) <0x00301>
  at System.Threading.Tasks.Task.SpinThenBlockingWait (int,System.Threading.CancellationToken) <0x000bd>
  at System.Threading.Tasks.Task.InternalWait (int,System.Threading.CancellationToken) <0x000af>
  at System.Threading.Tasks.Task.Wait (int,System.Threading.CancellationToken) <0x00069>
  at System.Threading.Tasks.Task.Wait () <0x00028>
  at NuGet.CommandLine.Command.Execute () <0x002f9>
  at NuGet.CommandLine.Program.MainCore (string,string[]) <0x008db>
  at NuGet.CommandLine.Program.Main (string[]) <0x00028>
  at (wrapper runtime-invoke) <Module>.runtime_invoke_int_object (object,intptr,intptr,intptr) <0xffffffff>

"<threadpool thread>"
```

**Comment by yishaigalatzer** [comment]
This does seem to be in mono proper
CC @mrward

**Comment by balneaves** [comment]
So this seems resolved in Mono 4.4.1

Although (on centos at least) there is a missing library bug which I had to fix with 
`sudo ln -s /usr/lib64/libMonoPosixHelper.so /usr/lib/libMonoPosixHelper.so`

**Comment by jainaashish** [comment]
It seems this has been fixed with latest release. If not, then please open another issue with all the details
